content/courses/advanced-ada/chapters/expressions.rst

```diff
@@ -551,10 +551,332 @@ won't trigger compilation errors. (Feel free to try this out!)
+
+
+    !show_reduction_expression.adb
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Reduction_Expression is
+       type Integer_Array is
+         array (Positive range <>) of Integer;
+
+       A : Integer_Array (1 .. 3);
+
+       type Integer_Accumulator is record
+          Value : Long_Integer;
+          Count : Integer;
+       end record;
+
+       function Accumulate
+         (Accumulator : Integer_Accumulator;
+          Value       : Integer)
+          return Integer_Accumulator is
+       begin
+          return (Value => Accumulator.Value
+                           + Long_Integer (Value),
+                  Count => Accumulator.Count + 1);
+       end Accumulate;
+
+       function Zero return Integer_Accumulator is
+         (Value => 0, Count => 0);
+
+       function Average (Acc : Integer_Accumulator)
+                         return Float is
+         (Float (Acc.Value) / Float (Acc.Count));
+
+       Acc : Integer_Accumulator;
+
+    begin
+       A := [2, 3, 4];
+
+       Acc := A'Reduce (Accumulate, Zero);
+       Put_Line ("Acc = "
+                 & Acc'Image);
+       Put_Line ("Avg = "
+                 & Average (Acc)'Image);
+    end Show_Reduction_Expression;
+
+In this example, we're using the :ada:`Integer_Accumulator` record type in our
+reducer |mdash| the :ada:`Accumulate` function. In this case, we're not only
+accumulating the values, but also counting the number of elements in the
+list. (Of course, we could have used :ada:`A'Length` for that as well.)
+
+Also, we're not limited to numeric types: we can also create a reducer using
+strings as the accumulator type. In fact, we can display the initial value and
+the elements of the list by using unbounded strings:
+
+.. code:: ada run_button manual_chop project=Courses.Advanced_Ada.Expressions.Reducer_String_Accumulator switches=Compiler(-gnatX)
+
+    !show_reduction_expression.adb
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    with Ada.Strings.Unbounded;
+    use  Ada.Strings.Unbounded;
+
+    procedure Show_Reduction_Expression is
+       type Integer_Array is
+         array (Positive range <>) of Integer;
+
+       A : Integer_Array (1 .. 3);
+
+       function Unbounded_String_List
+         (Accumulator : Unbounded_String;
+          Value       : Integer)
+              return Unbounded_String is
+       begin
+          return Accumulator
+                 & ", " & Value'Image;
+       end Unbounded_String_List;
+
+    begin
+       A := [2, 3, 4];
+
+       Put_Line ("A = "
+                 & A'Image);
+       Put_Line ("L = "
+                 & To_String (A'Reduce
+                   (Unbounded_String_List,
+                      To_Unbounded_String ("0"))));
+    end Show_Reduction_Expression;
+
+In this case, the "accumulator" is concatenating the initial value and
+individual values of the list into a string.
```

Comment:
That's pretty cool!

Reply:
As discussed separately, let's leave this topic (parallelism) out for the moment.